cpp/path_module/algorithm/path.hpp

```diff
@@ -44,64 +45,113 @@ struct LabelSets {
+  bool IsNotStartOrSupportsStartNode(bool is_start) const { return (config_.filter_start_node || !is_start); }
+  bool IsNotStartOrSupportsStartRel(bool is_start) const { return (config_.begin_sequence_at_start || !is_start); }
 
-void FilterLabel(const std::string_view label, const LabelSets &labelFilters, LabelBools &labelBools);
+  bool AreLabelsValid(const LabelBools &label_bools) const;
+  bool ContinueExpanding(const LabelBools &label_bools, size_t path_size) const;
 
-void ParseLabels(const mgp::List &list_of_labels, LabelSets &labelSets);
+  bool PathSizeOk(int64_t path_size) const;
+  bool PathTooBig(int64_t path_size) const;
+  bool Whitelisted(bool whitelisted) const;
 
-void ParseRelationships(const mgp::List &list_of_relationships, RelationshipSets &relationshipSets, bool &any_outgoing,
-                        bool &any_incoming);
-bool PathSizeOk(const int64_t path_size, const int64_t &max_hops, const int64_t &min_hops);
+  // methods for parsing config
+  void FilterLabelBoolStatus();
+  void FilterLabel(std::string_view label, LabelBools &label_bools) const;
+  void ParseLabels(const mgp::List &list_of_labels);
+  void ParseRelationships(const mgp::List &list_of_relationships);
 
-bool RelationshipAllowed(const std::string &rel_type, const RelationshipSets &relationshipSets, bool &any_outgoing,
-                         bool &any_incoming, bool outgoing);
+ private:
+  Config config_;
+};
 
-bool Whitelisted(const bool &whitelisted, const bool &whitelist_empty);
+class PathExpand {
+ public:
+  explicit PathExpand(PathHelper &&helper, const mgp::RecordFactory &record_factory, const mgp::Graph &graph)
+      : helper_(helper), record_factory_(record_factory), graph_(graph) {}
+
+  void ExpandPath(mgp::Path &path, const mgp::Relationship &relationship, int64_t path_size);
+  void ExpandFromRelationships(mgp::Path &path, mgp::Relationships relationships, bool outgoing, int64_t path_size,
+                               std::set<std::pair<std::string_view, int64_t>> &seen);
+  void StartAlgorithm(mgp::Node node);
+  void Parse(const mgp::Value &value);
+  void DFS(mgp::Path &path, int64_t path_size);
+  void RunAlgorithm();
+
+ private:
+  PathHelper helper_;
+  const mgp::RecordFactory &record_factory_;
+  const mgp::Graph &graph_;
+  std::unordered_set<int64_t> visited_;
+  std::unordered_set<mgp::Node> start_nodes_;
+};
 
-void PathDFS(mgp::Path path, std::unordered_set<mgp::Relationship> &relationships_set,
-             const mgp::RecordFactory &record_factory, int64_t path_size, const int64_t min_hops,
-             const int64_t max_hops, const LabelSets &labelFilters, const LabelBoolsStatus &labelStatus,
-             const RelationshipSets &relationshipSets, bool &any_outgoing, bool &any_incoming);
+class PathSubgraph {
+ public:
+  explicit PathSubgraph(PathHelper &&helper, const mgp::RecordFactory &record_factory, const mgp::Graph &graph)
+      : helper_(helper), record_factory_(record_factory), graph_(graph) {}
```

Comment:
You need to call on helper std::move, otherwise, it is copy constructed: https://godbolt.org/z/rxxKxnzjs

helper is still lvalue, you need to cast it with std::move to rvalue:https://godbolt.org/z/o4se6orET